### PR TITLE
[Merged by Bors] - feat: add basic API lemmas for `BddAbove (range f)` and `BddBelow (range f)`

### PR DESCRIPTION
**Pull request by dupuisf**
This PR adds some basic API lemmas for `BddAbove (range f)` (resp `BddBelow`). This is a ubiquitous side condition when working with `iInf`/`iSup` in conditionally complete lattices, so I think it's worth having.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Order/Bounds/Basic.lean [hunk 1]
```diff
@@ -1636,6 +1636,41 @@ theorem isGLB_prod [Preorder α] [Preorder β] {s : Set (α × β)} (p : α × 
...
+
+lemma BddBelow.range_comp {γ : Type*} [Preorder β] [Preorder γ] {f : α → β} {g : β → γ}
+    (hf : BddBelow (Set.range f)) (hg : Monotone g) : BddBelow (Set.range (fun x => g (f x))) := by
+  rw [bddBelow_def] at hf ⊢
+  obtain ⟨C, hC⟩ := hf
+  refine ⟨g C, fun y hy => ?_⟩
+  rw [Set.mem_range] at hy
+  obtain ⟨y', hy'⟩ := hy
+  calc y = g (f y') := hy'.symm
+      _ ≥ g C := hg <| hC (f y') <| Set.mem_range_self _
```
Comment (jcommelin) [inline Mathlib/Order/Bounds/Basic.lean:1672]: Can't this also be golfed using `OrderDual`?

Comment (dupuisf) [reply Mathlib/Order/Bounds/Basic.lean:1672]: Not easily. We need to take the dual of both `β` and `γ`, and I think the `Monotone g` hypothesis needs to be adapted. Maybe it's possible but it's easier to copy/paste.

**Review by ADedecker — COMMENTED** [review]
Could you also replace the uses of `Set.range` by `range` where the namespaces are already open?

File: Mathlib/Algebra/Bounds.lean [hunk 2]
```diff
@@ -135,6 +152,25 @@ theorem BddBelow.mul {s t : Set M} (hs : BddBelow s) (ht : BddBelow t) : BddBelo
 #align bdd_below.mul BddBelow.mul
 #align bdd_below.add BddBelow.add
 
+@[to_additive]
+lemma BddAbove.range_mul {α : Type*} {f g : α → M} (hf : BddAbove (Set.range f))
```
Comment (ADedecker) [inline Mathlib/Algebra/Bounds.lean:156]: Hopefully we should be able to make a proof of this analogous to the one I suggested for `inv` using lemmas added in #9690

File: Mathlib/Order/Bounds/Basic.lean [hunk 3]
```diff
@@ -1636,6 +1636,41 @@ theorem isGLB_prod [Preorder α] [Preorder β] {s : Set (α × β)} (p : α × 
...
+
+lemma BddAbove.range_comp {γ : Type*} [Preorder β] [Preorder γ] {f : α → β} {g : β → γ}
+    (hf : BddAbove (Set.range f)) (hg : Monotone g) : BddAbove (Set.range (fun x => g (f x))) := by
+  rw [bddAbove_def] at hf ⊢
+  obtain ⟨C, hC⟩ := hf
+  refine ⟨g C, fun y hy => ?_⟩
+  rw [Set.mem_range] at hy
+  obtain ⟨y', hy'⟩ := hy
+  calc y = g (f y') := hy'.symm
+      _ ≤ g C := hg <| hC (f y') <| Set.mem_range_self _
```
[the remark this answers is not in the document]
Comment (dupuisf) [reply Mathlib/Order/Bounds/Basic.lean]: Yes, the version with comp notation is more annoying to use in practice. Here's an example:
```
example : BddBelow (range (fun (n : ℤ) => Int.natAbs (n + 5))) := by
  refine BddBelow.range_comp ?_ ?_
  sorry
```
Here, the refine call fails for `range_comp` defined as `BddBelow (range (g ∘ f))` but succeeds for `BddBelow (range (fun x => g (f x)))`. I see quite a few lemmas in mathlib with this problem, and having to use `change` to apply them gets tedious. I'll change it back if you don't mind.

File: Mathlib/Algebra/Bounds.lean [hunk 4]
```diff
@@ -88,6 +88,23 @@ theorem IsLUB.inv (h : IsLUB s a) : IsGLB s⁻¹ a⁻¹ :=
 #align is_lub.inv IsLUB.inv
 #align is_lub.neg IsLUB.neg
 
+@[to_additive]
+lemma BddBelow.bddAbove_range_inv  {α : Type*} {f : α → G} (hf : BddBelow (Set.range f)) :
```
[the remark this answers is not in the document]
Comment (dupuisf) [reply Mathlib/Algebra/Bounds.lean:92]: Fair enough.

File: Mathlib/Algebra/Bounds.lean (hunk 2 above)
Comment (dupuisf) [reply Mathlib/Algebra/Bounds.lean]: Good idea; I'll make the change as soon as it's done merging.

File: Mathlib/Order/Bounds/Basic.lean (hunk 3 above)
[the remark this answers is not in the document]
Comment (jcommelin) [reply Mathlib/Order/Bounds/Basic.lean]: I guess you can still use Anatole's proof after `show {the statement using \circ}`.

[the remark this answers is not in the document]
Comment (dupuisf) [reply Mathlib/Order/Bounds/Basic.lean]: Yes, that's what I did.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Comment by jcommelin** [comment]
bors r-

Sorry, I thought you were done...

bors d+

File: Mathlib/Algebra/Bounds.lean (hunk 2 above)
Comment (dupuisf) [reply Mathlib/Algebra/Bounds.lean]: It's done now.

**Comment by dupuisf** [comment]
bors merge

**Pull request closed without merging**